ledger-core/v2/vanilla/atomickit/oncestate.go

```
@@ -0,0 +1,176 @@
+}
+
+func (p *OnceFlag) DoSet(f func()) bool {
+	if !atomic.CompareAndSwapInt32(&p.done, 0, -1) {
+		return false
+	}
+	p.doSlow(f)
+	return true
+}
+
+func (p *OnceFlag) DoSpin(f func()) bool {
+	if !atomic.CompareAndSwapInt32(&p.done, 0, -1) {
+		for !p.IsSet() {
+			runtime.Gosched()
+		}
+		return false
+	}
+	p.doSlow(f)
+	return true
+}
+
+func (p *OnceFlag) doSlow(f func()) {
+	defer atomic.StoreInt32(&p.done, 1)
+	f()
+}
+
+/***********************************************************/
+
+type StartStopFlag struct {
+	done int32
+}
+
+func (p *StartStopFlag) IsActive() bool {
+	return atomic.LoadInt32(&p.done) == 1
+}
+
+func (p *StartStopFlag) IsStarting() bool {
+	return atomic.LoadInt32(&p.done) < 0
+}
+
+func (p *StartStopFlag) WasStarted() bool {
+	return atomic.LoadInt32(&p.done) != 0
+}
+
+func (p *StartStopFlag) WasStopped() bool {
+	return atomic.LoadInt32(&p.done) >= 2
+}
+
+func (p *StartStopFlag) IsStopping() bool {
+	return atomic.LoadInt32(&p.done) == 2
+}
+
+func (p *StartStopFlag) IsStopped() bool {
+	return atomic.LoadInt32(&p.done) == 3
+}
+
+func (p *StartStopFlag) Status() (active, wasStarted bool) {
+	n := atomic.LoadInt32(&p.done)
+	return n == 1, n != 0
+}
+
+func (p *StartStopFlag) DoStart(f func()) bool {
+	if !atomic.CompareAndSwapInt32(&p.done, 0, -1) {
+		return false
+	}
+	p.doSlow(f, 1)
+	return true
+}
+
+func (p *StartStopFlag) DoStop(f func()) bool {
+	if !atomic.CompareAndSwapInt32(&p.done, 1, 2) {
+		return false
+	}
+	p.doSlow(f, 3)
+	return true
+}
+
+func (p *StartStopFlag) DoDiscard(discardFn, stopFn func()) bool {
+	for {
+		switch atomic.LoadInt32(&p.done) {
+		case 0:
+			if atomic.CompareAndSwapInt32(&p.done, 0, 2) {
+				p.doSlow(discardFn, 3)
+				return true
+			}
+		case 1:
+			return p.DoStop(stopFn)
+		default:
+			return false
+		}
+	}
+}
+
+func (p *StartStopFlag) Start() bool {
+	return !atomic.CompareAndSwapInt32(&p.done, 0, 1)
+}
+
+func (p *StartStopFlag) Stop(f func()) bool {
```

Why is it accepting `f`, if it's silently ignores it?

---

stupid err